src/tokens/ERC1155/ERC1155Sale.sol

```diff
@@ -0,0 +1,296 @@
+    bytes32 public constant ROYALTY_ADMIN_ROLE = keccak256("ROYALTY_ADMIN_ROLE");
+
+    bytes4 private constant ERC20_TRANSFERFROM_SELECTOR =
+        bytes4(keccak256(bytes("transferFrom(address,address,uint256)")));
+
+    bool private _initialized;
+
+    // ERC20 token address for payment. address(0) indicated payment in ETH.
+    address private _paymentToken;
+
+    SaleDetails private _globalSaleDetails;
+    mapping(uint256 => SaleDetails) private _tokenSaleDetails;
+
+    /**
+     * Initialize with empty values.
+     * @dev These are overridden by initialize().
+     */
+    constructor() ERC1155Metadata("", "") {}
+
+    /**
+     * Initialize the contract.
+     * @param _owner Owner address.
+     * @param _name Token name.
+     * @param _baseURI Base URI for token metadata.
+     * @dev This should be called immediately after deployment.
+     */
+    function initialize(address _owner, string memory _name, string memory _baseURI) public {
+        if (_initialized) {
+            revert InvalidInitialization();
+        }
+        _initialized = true;
+        name = _name;
+        baseURI = _baseURI;
+        _setupRole(DEFAULT_ADMIN_ROLE, _owner);
+        _setupRole(MINT_ADMIN_ROLE, _owner);
+        _setupRole(ROYALTY_ADMIN_ROLE, _owner);
+    }
+
+    /**
+     * Checks if the current block.timestamp is out of the give timestamp range.
+     * @param startTime Earliest acceptable timestamp (inclusive).
+     * @param endTime Latest acceptable timestamp (exclusive).
+     * @dev A zero endTime value is always considered out of bounds.
+     */
+    function blockTimeOutOfBounds(uint256 startTime, uint256 endTime) private view returns (bool) {
+        // 0 end time indicates inactive sale.
+        return endTime == 0 || block.timestamp < startTime || block.timestamp >= endTime;
+    }
+
+    /**
+     * Checks the sale is active and takes payment.
+     * @param _tokenIds Token IDs to mint.
+     * @param _amounts Amounts of tokens to mint.
+     */
+    function payForActiveMint(uint256[] memory _tokenIds, uint256[] memory _amounts) private {
+        uint256 totalAmount;
+        uint256 totalCost;
+        SaleDetails memory gSaleDetails = _globalSaleDetails;
+        bool globalSaleInactive = blockTimeOutOfBounds(gSaleDetails.startTime, gSaleDetails.endTime);
+        for (uint256 i; i < _tokenIds.length; i++) {
+            // Active sale test
+            SaleDetails memory saleDetails = _tokenSaleDetails[_tokenIds[i]];
+            bool tokenSaleInactive = blockTimeOutOfBounds(saleDetails.startTime, saleDetails.endTime);
+            if (tokenSaleInactive) {
+                // Prefer token sale
+                if (globalSaleInactive) {
+                    // Both sales inactive
+                    revert SaleInactive(_tokenIds[i]);
+                }
+                // Use global sale details
+                totalCost += gSaleDetails.cost * _amounts[i];
```

Comment:
This is a nitpick, but maybe it would be nice to avoid duplicating `total += cost * amount`, these conditions may define a `cost` variable that is later used. This is a minor thing so feel free to ignore if you don't agree.